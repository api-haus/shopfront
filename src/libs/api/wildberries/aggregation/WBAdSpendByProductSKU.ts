import { batchSplit } from '../../../batching/batchSplit.js';
import { DatesBetween } from '../../../date/DatesBetween.js';
import { Pause } from '../../../Pause.js';
import type {
  CachedWildberriesAdvertAPI,
} from '../accessors/CachedWildberriesAdvertAPI.js';
import type { WildberriesAdvertAPI } from '../clients/WildberriesAdvertAPI.js';
import { idDatesArg } from '../params/IDAndDatesArg.js';
import type { IWBPromotionFullStats } from '../types/IWBPromotionFullStats.js';

export class WBAdSpendByProductSKU {
  constructor(protected wbAdvertsAPI: WildberriesAdvertAPI, protected cachedWbAdvertsAPI: CachedWildberriesAdvertAPI) {
  }

  async adSpendBySKU(dateFrom: Date, dateTo: Date) {
    const { wbAdvertsAPI, cachedWbAdvertsAPI } = this;

    const { adverts } = await wbAdvertsAPI.groupedPromotionCounts();
    const advertIDs = adverts.flatMap((av) => {
      return av.advert_list.map(av => av.advertId);
    });

    const datesBetween = DatesBetween(
      dateFrom,
      dateTo,
    );
    const batchesBy5Days = batchSplit(
      datesBetween,
      5,
    );
    const batchesBy100AdIds = batchSplit(
      advertIDs,
      100,
    );

    const batchesOf100IdsAnd5Dates = batchesBy100AdIds.flatMap((hundredIds) => {
      return hundredIds.map((id) => {
        return batchesBy5Days.map(fiveDates => idDatesArg(
          id,
          fiveDates,
        ));
      });
    });

    const allPromotionStats: IWBPromotionFullStats[] = [];

    for (const batchArgs of batchesOf100IdsAnd5Dates) {
      const weight = batchArgs.reduce(
        (a, v) => a + v.dates.length,
        0,
      );

      const batch = await cachedWbAdvertsAPI.promotionFullStats(batchArgs);
      allPromotionStats.push(...batch);

      await Pause(weight * 10_000);
    }

    return allPromotionStats;
  }
}
