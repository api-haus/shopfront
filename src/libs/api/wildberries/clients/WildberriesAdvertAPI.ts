import { constants } from 'node:http2';

import Bottleneck from 'bottleneck';
import { getGlobalDispatcher, interceptors } from 'undici';

import config from '../../../../config/index.js';
import { TokenizedAPIClient } from '../../clients/TokenizedAPIClient.js';
import { retryHandler } from '../../common/RetryHandler.js';
import type { IIdAndDates } from '../params/IDAndDatesArg.js';
import type { IPromotionAdvertsInfo } from '../types/IPromotionAdvertsInfo.js';
import type { IWBAdvertList } from '../types/IWBAdvertList.js';
import type { IWBPromotionFullStats } from '../types/IWBPromotionFullStats.js';

export const CPrefixAdvertAPIV1 = 'https://advert-api.wildberries.ru/adv/v1';
export const CPrefixAdvertAPIV2 = 'https://advert-api.wildberries.ru/adv/v2';

export class WildberriesAdvertAPI extends TokenizedAPIClient {
  bottleneck = new Bottleneck({
    maxConcurrent: 1,
    minTime: 5000,
  });

  dispatcher = getGlobalDispatcher()
    .compose(interceptors.retry({
      ...config.wildberries.api.retryOptions,
      retry: retryHandler,
      errorCodes: [
        'ECONNRESET',
        'ECONNREFUSED',
        'ENOTFOUND',
        'ENETDOWN',
        'ENETUNREACH',
        'EHOSTDOWN',
        'EHOSTUNREACH',
        'EPIPE',
        'UND_ERR_SOCKET',
        'UND_ERR_REQ_CONTENT_LENGTH_MISMATCH',
      ],
      methods: [
        'GET',
        'HEAD',
        'POST',
        'OPTIONS',
        'PUT',
        'DELETE',
        'TRACE',
      ],
      statusCodes: [
        500,
        502,
        503,
        504,
        429,
      ],
    }));

  async groupedPromotionCounts(): Promise<{ adverts: IWBAdvertList[] }> {
    return this.rateLimitedCall<{ adverts: IWBAdvertList[] }>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV1}/promotion/count`,
      {
        method: 'GET',
      },
    );
  }

  async promotionFullStats(idsAndDates: IIdAndDates[]): Promise<IWBPromotionFullStats[]> {
    return this.rateLimitedCall<IWBPromotionFullStats[]>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV2}/fullstats`,
      {
        method: 'POST',
        headers: {
          [constants.HTTP2_HEADER_CONTENT_TYPE]: 'application/json',
        },
        body: JSON.stringify(idsAndDates),
      },
    ).catch((err: unknown) => {
      if ((err as Error).message.includes('нет кампаний')) {
        return [];
      }
      throw err;
    });
  }

  async promotionAdvertsInfo(advertIds: number[]): Promise<IPromotionAdvertsInfo> {
    return this.rateLimitedCall<IPromotionAdvertsInfo>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV1}/promotion/adverts`,
      {
        method: 'POST',
        headers: {
          [constants.HTTP2_HEADER_CONTENT_TYPE]: 'application/json',
        },
        body: JSON.stringify(advertIds),
      },
    );
  }
}
