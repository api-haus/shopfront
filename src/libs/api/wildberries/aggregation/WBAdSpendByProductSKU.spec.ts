import { expect, test } from '@jest/globals';
import { setupRecorder } from 'nock-record';

import { testMongoClient } from '../../../test/fixtures/testMongoClient.js';
import {
  testWbTokenProvider,
} from '../../../test/fixtures/testWbTokenProvider.js';
import {
  CachedWildberriesAdvertAPI,
} from '../accessors/CachedWildberriesAdvertAPI.js';
import { WildberriesAdvertAPI } from '../clients/WildberriesAdvertAPI.js';
import { WBAdSpendByProductSKU } from './WBAdSpendByProductSKU.js';

const record = setupRecorder();

test(
  'same dates',
  async () => {
    const { completeRecording, assertScopesFinished } = await record('__WBAdSpendByProductSKU');

    const wbAdvertsAPI = new WildberriesAdvertAPI(testWbTokenProvider);
    const cachedWildberriesAdvertAPI = new CachedWildberriesAdvertAPI(
      wbAdvertsAPI,
      testMongoClient,
    );
    const wbAdSpendByProductSKU = new WBAdSpendByProductSKU(
      wbAdvertsAPI,
      cachedWildberriesAdvertAPI,
    );

    const dateFrom = new Date('2024-06-21');
    const dateTo = new Date('2024-06-21');

    const totalRows = await wbAdSpendByProductSKU.adSpendBySKU(
      dateFrom,
      dateTo,
    );

    completeRecording();
    assertScopesFinished();

    expect(totalRows).toMatchSnapshot();
  },
  2_000_000,
);

test(
  'WBAdSpendByProductSKU',
  async () => {
    const { completeRecording, assertScopesFinished } = await record('__WBAdSpendByProductSKU');

    const wbAdvertsAPI = new WildberriesAdvertAPI(testWbTokenProvider);
    const cachedWildberriesAdvertAPI = new CachedWildberriesAdvertAPI(
      wbAdvertsAPI,
      testMongoClient,
    );
    const wbAdSpendByProductSKU = new WBAdSpendByProductSKU(
      wbAdvertsAPI,
      cachedWildberriesAdvertAPI,
    );

    const dateFrom = new Date('2024-06-21');
    const dateTo = new Date('2024-06-22');

    const totalRows = await wbAdSpendByProductSKU.adSpendBySKU(
      dateFrom,
      dateTo,
    );

    completeRecording();
    assertScopesFinished();

    expect(totalRows).toMatchSnapshot();
  },
  2_000_000,
);
