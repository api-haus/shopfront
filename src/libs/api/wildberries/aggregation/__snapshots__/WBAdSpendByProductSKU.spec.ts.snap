// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`same dates 1`] = `[]`;
